src/measure_theory/function/conditional_expectation.lean

```diff
@@ -1266,6 +1319,314 @@ lemma condexp_ind_disjoint_union (hs : measurable_set s) (ht : measurable_set t)
+end
+
+lemma integrable_condexp_L1 (f : α → F') : integrable (condexp_L1 hm μ f) μ :=
+L1.integrable_coe_fn _
+
+lemma set_integral_condexp_L1 (hf : integrable f μ) (hs : measurable_set[m] s) :
+  ∫ x in s, condexp_L1 hm μ f x ∂μ = ∫ x in s, f x ∂μ :=
+begin
+  simp_rw condexp_L1_eq hf,
+  rw set_integral_condexp_L1_clm (hf.to_L1 f) hs,
+  exact set_integral_congr_ae (hm s hs) ((hf.coe_fn_to_L1).mono (λ x hx hxs, hx)),
+end
+
+lemma condexp_L1_add (hf : integrable f μ) (hg : integrable g μ) :
+  condexp_L1 hm μ (f + g) = condexp_L1 hm μ f + condexp_L1 hm μ g :=
+calc condexp_L1 hm μ (f + g) = condexp_L1_clm hm μ ((hf.add hg).to_L1 (f + g)) :
+  condexp_L1_eq (hf.add hg)
+... = condexp_L1_clm hm μ (hf.to_L1 f + hg.to_L1 g) : by rw integrable.to_L1_add _ _ hf hg
+... = condexp_L1_clm hm μ (hf.to_L1 f) + condexp_L1_clm hm μ (hg.to_L1 g) :
+  (condexp_L1_clm hm μ).map_add _ _
+... = condexp_L1 hm μ f + condexp_L1 hm μ g :
+  by rw [condexp_L1_eq hf, condexp_L1_eq hg]
+
+lemma condexp_L1_neg (f : α → F') :
+  condexp_L1 hm μ (-f) = - condexp_L1 hm μ f :=
+begin
+  by_cases hf : integrable f μ,
+  { calc condexp_L1 hm μ (-f) = condexp_L1_clm hm μ (hf.neg.to_L1 (-f)) : condexp_L1_eq hf.neg
+    ... = condexp_L1_clm hm μ (- hf.to_L1 f) : by rw integrable.to_L1_neg _ hf
+    ... = - condexp_L1_clm hm μ (hf.to_L1 f) : (condexp_L1_clm hm μ).map_neg _
+    ... = - condexp_L1 hm μ f : by rw condexp_L1_eq hf, },
+  { rw [condexp_L1_undef hf, condexp_L1_undef (mt integrable_neg_iff.mp hf), neg_zero], },
+end
+
+lemma condexp_L1_smul (c : 𝕜) (f : α → F') :
+  condexp_L1 hm μ (c • f) = c • condexp_L1 hm μ f :=
+begin
+  by_cases hf : integrable f μ,
+  { calc condexp_L1 hm μ (c • f) = condexp_L1_clm hm μ ((hf.smul c).to_L1 (c • f)) :
+      condexp_L1_eq (hf.smul c)
+    ... = condexp_L1_clm hm μ (c • hf.to_L1 f) : by rw integrable.to_L1_smul' _ hf c
+    ... = c • condexp_L1_clm hm μ (hf.to_L1 f) : condexp_L1_clm_smul c (hf.to_L1 f)
+    ... = c • condexp_L1 hm μ f : by rw condexp_L1_eq hf, },
+  { by_cases hc : c = 0,
+    { rw [hc, zero_smul, zero_smul, condexp_L1_zero], },
+    rw [condexp_L1_undef hf, condexp_L1_undef (mt (integrable_smul_iff hc f).mp hf), smul_zero], },
+end
+
+lemma condexp_L1_sub (hf : integrable f μ) (hg : integrable g μ) :
+  condexp_L1 hm μ (f - g) = condexp_L1 hm μ f - condexp_L1 hm μ g :=
+by rw [sub_eq_add_neg, sub_eq_add_neg, condexp_L1_add hf hg.neg, condexp_L1_neg g]
+
+end condexp_L1
+
+section condexp
+
+/-! # Conditional expectation of a function -/
+
+open_locale classical
+
+local attribute [instance] fact_one_le_one_ennreal
+
+variables {𝕜} {m m0 : measurable_space α} {μ : measure α} [borel_space 𝕜] [is_scalar_tower ℝ 𝕜 F']
+  {hm : m ≤ m0} [sigma_finite (μ.trim hm)] {f g : α → F'} {s : set α}
+
+/-- Conditional expectation of a function. Its value is 0 if the function is not integrable. -/
+def condexp (hm : m ≤ m0) (μ : measure α) [sigma_finite (μ.trim hm)] (f : α → F') : α → F' :=
```

Comment:
What about marking this as `irreducible`, to make sure Lean doesn't get lost trying to unfold too many layers?